File: SoftUni - C# OOP/Encapsulation/Encapsulation - Lab/02. Salary/Person.cs
﻿using System;
using System.Collections.Generic;
using System.Text;

namespace PersonsInfo
{
    public class Person
    {         
        public string FirstName 
        {   
            get
            {
                return this.firstName;
            }
            private set
            {           
                this.firstName = value;
            } 
        }
        public string LastName
        {
            get
            {
                return this.lastName;
            }
            private set
            {             
                this.lastName = value;
            }
        }
        public int Age
        {
            get
            {
                return this.age;
            }
            private set
            {
                this.age = value;
            }
        }
        public decimal Salary
        {
            get
            {
                return this.salaly;
            }
            private set
            {
                this.salaly = value;
            }
        }


        public Person(string firstName, string lastName, int age,decimal salary)
        {
            this.FirstName = firstName;
            this.LastName = lastName;
            this.Age = age;
            this.Salary = salary;
        }
     
        public void IncreaseSalary(decimal percentage)
        {
            if (this.Age > 30)
            {
                this.Salary += this.Salary * percentage / 100;
            }
            else
            {
                this.Salary += this.Salary * percentage / 200;
            }
        }
        public override string ToString()
        {
            return $"{this.FirstName} {this.LastName} receives {this.Salary:F2} leva.";
        }

    }
}

File: SoftUni - Java OOP/Encapsulation/Encapsulation - Lab/_03_Validation_Data/Person.java
package _03_Validation_Data;

public class Person {
    private String firstName;
    private String lastName;
    private int age;
    private double salary;

    public Person(String firstName, String lastName, int age, double salary) {

        // business logic → not good in the constructor
        /*if(firstName.length() < 3) {
            throw new IllegalStateException();
        }
         */
        setFirstName(firstName);
        setLastName(lastName);
        setAge(age);
        setSalary(salary);
    }
    public String toString() {
        return String.format("%s %s gets %f leva.", this.firstName, this.lastName, this.salary);
    }

    public void increaseSalary(double bonus) {
        if(this.age < 30) {
            this.salary += this.salary  * (bonus / 200);
        } else {
            this.salary += this.salary  * (bonus / 100);
        }
    }

    public String getFirstName() {
        return this.firstName;
    }

    public String getLastName() {
        return this.lastName;
    }
    public int getAge() {
        return this.age;
    }
    public double getSalary() {
        return salary;
    }
    public void setFirstName(String firstName) {
        if(firstName.length() < 3) {
            throw new IllegalArgumentException("First name cannot be less than 3 symbols");
        }
        this.firstName = firstName;
    }
    public void setLastName(String lastName) {
        if(lastName.length() < 3) {
            throw new IllegalArgumentException("Last name cannot be less than 3 symbols");
        }
        this.lastName = lastName;
    }
    public void setAge(int age) {
        if(age <= 0) {
            throw new IllegalArgumentException("Age cannot be zero or negative integer");
        }
        this.age = age;
    }
    public void setSalary(double salary) {
        if(salary < 460) {
            throw new IllegalArgumentException("Salary cannot be less than 460 leva");
        }
        this.salary = salary;
    }
}
